File: python/Tree/BFS/invert_binary_tree.py
'''

175. Invert Binary Tree
Invert a binary tree.

Example
  1         1
 / \       / \
2   3  => 3   2
   /       \
  4         4
Challenge
Do it in recursion is acceptable, can you do it without recursion?

'''
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param root: a TreeNode, the root of the binary tree
    @return: nothing
    """
    def invertBinaryTree(self, root):
        # write your code here
        stack = []
        stack.append(root)

        while stack:
            cur = stack.pop()
            cur.left, cur.right = cur.right, cur.left

            if cur.left:
                stack.append(cur.left)
            if cur.right:
                stack.append(cur.right)

File: Tree/recursion/invert-binary-tree/invert-binary-tree.ts
/**
 * @param root: a TreeNode, the root of the binary tree
 * @return:
 */
const invertBinaryTree = function (root) {
  // write your code here

  if (root === null) {
    return null;
  }

  let temp = root.left;
  root.left = root.right;
  root.right = temp;

  invertBinaryTree(root.left);
  invertBinaryTree(root.right);
};
